### 1.1. Подготовка и создание данных — 4 балла

1. Импортируйте NumPy и задайте фиксированный seed для генератора случайных чисел.
2. Создайте `area_array` — 200 значений площади квартир в диапазоне от 25 до 150 м². Значения могут быть целыми или вещественными.
3. Создайте `rooms_array` — 200 целых значений количества комнат от 1 до 5 включительно.
4. Создайте `price_array` — 200 значений стоимости квартир в диапазоне от 15 до 120 млн тенге.
5. Убедитесь, что все три массива имеют одинаковое количество элементов и соответствуют одним и тем же 200 квартирам.
6. Для каждого массива выведите `shape`, `ndim`, `size` и `dtype`.

> **Подсказка:** Для генерации можно использовать `np.random.randint()`, `np.random.uniform()` или методы объекта `rng`. Значения `price_array` допускается округлить до 1–2 знаков.

**Что должно быть в результате:**
- Три массива длиной 200.
- Вывод основных атрибутов каждого массива.
- Короткое пояснение, что означают `shape`, `ndim`, `size` и `dtype`.

In [1]:
import numpy as np

In [2]:
rng = np.random.default_rng(42)
n = 200
area_array = np.round(rng.uniform(25, 150, size=n),2)
rooms_array = rng.integers(1, 6, size=n)
price_array = np.round(rng.uniform(15, 120, size=n),2)
print(len(area_array), len(rooms_array), len(price_array))

200 200 200


In [3]:
print(f'area_array: shape: {area_array.shape}, ndim: {area_array.ndim} size: {area_array.size} dtype: {area_array.dtype}')
print(f'rooms_array: shape: {rooms_array.shape}, ndim: {rooms_array.ndim} size: {rooms_array.size} dtype: {rooms_array.dtype}')
print(f'price_array: shape: {price_array.shape}, ndim: {price_array.ndim} size: {price_array.size} dtype: {price_array.dtype}')

area_array: shape: (200,), ndim: 1 size: 200 dtype: float64
rooms_array: shape: (200,), ndim: 1 size: 200 dtype: int64
price_array: shape: (200,), ndim: 1 size: 200 dtype: float64


shape - показывает количество элементов в dimensions
ndim - показвает количество dimensions 
size - общее количество элементов
dtype - тип данных в array

### 1.2. Индексация, срезы и булева фильтрация — 5 баллов

1. Выведите первые 10 и последние 10 элементов каждого массива.
2. Выберите и выведите площади квартир с индексами от 20 до 39 включительно.
3. Найдите количество квартир площадью более 70 м².
4. Найдите квартиры со стоимостью выше среднего значения `price_array`. Выведите их количество и первые 10 найденных цен.
5. Создайте одну составную булеву маску для квартир, у которых `area > 70` и `rooms >= 3`. Выведите количество таких квартир.
6. Усложните условие: найдите квартиры с 2–4 комнатами, площадью 60–100 м² и стоимостью не выше 80 млн тенге. Выведите индексы первых 10 подходящих квартир.

> **Подсказка:** Для составных условий используйте `&`, `|` и круглые скобки. Не используйте Python `and`/`or` для NumPy-массивов.

**Что должно быть в результате:**
- Примеры обычной индексации и slicing.
- Не менее трёх булевых масок.
- Количество объектов, удовлетворяющих каждому условию.


In [4]:
print(f'area_arrsy \nfisrt 10: {area_array[:10]}, \nlast 10: {area_array[n-10:]}\n')
print(f'rooms_arrsy \nfisrt 10: {rooms_array[:10]}, \nlast 10: {rooms_array[n-10:]}\n')
print(f'price_arrsy \nfisrt 10: {price_array[:10]}, \nlast 10: {price_array[n-10:]}\n')

area_arrsy 
fisrt 10: [121.74  79.86 132.32 112.17  36.77 146.95 120.14 123.26  41.01  81.3 ], 
last 10: [142.1   55.12  40.34 128.89  44.16  47.41  99.92 134.32  49.55  63.79]

rooms_arrsy 
fisrt 10: [5 4 2 5 3 3 3 1 5 1], 
last 10: [5 5 2 5 2 5 3 4 4 4]

price_arrsy 
fisrt 10: [ 96.79  29.13  71.29  68.99 105.05  63.59  55.43  82.15  42.98  29.68], 
last 10: [88.71 83.03 83.51 48.2  97.68 72.66 60.3  80.73 52.87 68.84]



In [5]:
print(f'area_arrsy indexes 20-39 {area_array[20:40]}')

area_arrsy indexes 20-39 [119.76  69.32 146.34 136.64 122.3   49.33  83.34  30.48  44.29 110.38
 118.1  145.94  65.73  71.31  83.69  48.68  41.24  84.46  53.36 108.73]


In [6]:
print(f'area_arrays > 70 m2  -> {len(area_array[area_array>70])}')

area_arrays > 70 m2  -> 128


In [7]:
price_avg = round(price_array.mean(), 2)
price_higher_avg = price_array[price_array > price_avg]
print(f'prices higher than avg qty: {len(price_higher_avg)}\nfirst 10 prices: {price_higher_avg[:10]}')

prices higher than avg qty: 104
first 10 prices: [ 96.79  71.29  68.99 105.05  82.15  87.   114.63 111.22  71.22 104.1 ]


In [8]:
mask = (area_array > 70) & (rooms_array >= 3)
print(f'arrea > 70 and rooms >= 3 qty: {len(rooms_array[mask])}')

arrea > 70 and rooms >= 3 qty: 78


In [9]:
mask = ((area_array >= 60) & 
        (area_array <= 100) &  
        (rooms_array >= 2) & 
        (rooms_array <= 4) & (price_array<=80))

print(f'first 10 indices under the condition: {np.where(mask)[0][:10]}')

print('Количество:', np.count_nonzero(mask))

first 10 indices under the condition: [  1  21  34  43  62  75  78  92  94 106]
Количество: 20


### 1.3. Описательная статистика — 6 баллов
> **Подсказка:** Понадобятся `np.mean`, `np.median`, `np.min`, `np.max`, `np.std`, `np.var`, `np.percentile`, `np.argmax`.

1. Для `area_array` и `price_array` вычислите mean, median, min, max, standard deviation и variance.

In [10]:
print(f'area_array mean: {np.mean(area_array):.2f}, median: {np.median(area_array):.2f}, min: {np.min(area_array):.2f}, max: {np.max(area_array):.2f}, std: {np.std(area_array):.2f}, var: {np.var(area_array):.2f}')
print()
print(f'price_array mean: {np.mean(price_array):.2f}, median: {np.median(price_array):.2f}, min: {np.min(price_array):.2f}, max: {np.max(price_array):.2f}, std: {np.std(price_array):.2f}, var: {np.var(price_array):.2f}')

area_array mean: 86.73, median: 84.88, min: 25.92, max: 149.05, std: 35.20, var: 1239.08

price_array mean: 68.47, median: 71.08, min: 15.57, max: 119.91, std: 29.49, var: 869.69


2. Для обоих массивов вычислите перцентили 10%, 25%, 50%, 75%, 90% и 99%.

In [11]:
print(f'area_array percentile \n10: {np.percentile(area_array, 10):.2f},\n25: {np.percentile(area_array, 25):.2f},\n50: {np.percentile(area_array, 50):.2f},\n75: {np.percentile(area_array, 75):.2f},\n90: {np.percentile(area_array, 90):.2f},\n99: {np.percentile(area_array, 99):.2f}')
print()
print(f'price_array percentile \n10: {np.percentile(price_array, 10):.2f},\n25: {np.percentile(price_array, 25):.2f},\n50: {np.percentile(price_array, 50):.2f},\n75: {np.percentile(price_array, 75):.2f},\n90: {np.percentile(price_array, 90):.2f},\n99: {np.percentile(price_array, 99):.2f}')

area_array percentile 
10: 38.56,
25: 56.53,
50: 84.88,
75: 118.41,
90: 136.36,
99: 146.35

price_array percentile 
10: 27.51,
25: 43.42,
50: 71.08,
75: 92.43,
90: 109.20,
99: 118.61


3. Определите индекс самой дорогой квартиры. По этому индексу выведите её price, area и rooms.

In [12]:
most_expensive_indx = price_array.argmax()

print(f'Most expensive apartment: price: {price_array[most_expensive_indx]:.2f}, area: {area_array[most_expensive_indx]:.2f}, rooms: {rooms_array[most_expensive_indx]}')

Most expensive apartment: price: 119.91, area: 137.83, rooms: 2


4. Определите индекс квартиры с максимальной площадью и выведите все её характеристики.


In [13]:
biggest_indx = np.argmax(area_array)
print(f'The biggest apartment: price: {price_array[biggest_indx]:.2f}, area: {area_array[biggest_indx]:.2f}, rooms: {rooms_array[biggest_indx]}')

The biggest apartment: price: 79.15, area: 149.05, rooms: 3


5. Посчитайте, сколько квартир имеют цену ниже медианы и сколько — площадь выше среднего значения.


In [14]:
price_median = np.median(price_array)
arrea_mean = np.mean(area_array)
price_lower_median = price_array[price_array < price_median]
arrea_bigger_mean = area_array[area_array>arrea_mean]
print(f'QTY of price < median price: {price_lower_median.size}')
print(f'QTY of arrea > mean arrea: {arrea_bigger_mean.size}')

QTY of price < median price: 100
QTY of arrea > mean arrea: 97


6. Сравните mean и median цены в 1–2 предложениях. Укажите, что может означать заметная разница между ними.


In [15]:
print(f'avg price {price_avg} and median price {price_median}')

avg price 68.47 and median price 71.08


Разница не такая большая. Возможно, квартиры с низкими ценами немного тянут среднее вниз

### 1.4. Векторизованные вычисления и новые признаки — 6 баллов

1. Создайте `price_per_m2_array` по формуле `price_array / area_array`. Цикл `for` для этого расчёта использовать нельзя.

Цена за 1 м² рассчитана в тенге/м²: стоимость квартиры в млн тенге умножена на 1 000 000 и разделена на площадь.

In [16]:
price_per_m2_array = np.round(price_array / area_array * 1000000, 2)
print(price_per_m2_array.size)

200


2. Найдите среднюю, медианную, минимальную и максимальную цену за 1 м².
3. Определите индекс квартиры с максимальной ценой за 1 м² и выведите её исходные характеристики.
4. Посчитайте количество квартир, у которых цена за 1 м² выше среднего значения.
5. Создайте условную категорию `expensive_flag` как булев массив: `True`, если цена квартиры выше 75-го перцентиля `price_array`, иначе `False`.
6. Проверьте, сколько элементов `True` и `False` содержит `expensive_flag`.

In [17]:
print('\n2')
price_per_m2_array_avg = np.mean(price_per_m2_array)
price_per_m2_array_median = np.median(price_per_m2_array)
price_per_m2_array_min = np.min(price_per_m2_array)
price_per_m2_array_max = np.max(price_per_m2_array)
print(f'price_per_m2_array mean: {price_per_m2_array_avg:.2f}, median: {price_per_m2_array_median:.2f}, min: {price_per_m2_array_min:.2f}, max: {price_per_m2_array_max:.2f}')

print('\n3')
price_per_m2_array_argmax = np.argmax(price_per_m2_array)
print(f'Характеристика квартиры у которой max price per m2: price: {price_array[price_per_m2_array_argmax]:.2f}, area: {area_array[price_per_m2_array_argmax]:.2f}, rooms: {rooms_array[price_per_m2_array_argmax]}')

print('\n4')
price_per_m2_array_bigger_avg = price_per_m2_array[price_per_m2_array > price_per_m2_array_avg]

print(f'QTY of apartmanets vy m2 > avg: {price_per_m2_array_bigger_avg.size}')

print('\n5and6')
expensive_flag = price_array > np.percentile(price_array, 75)
print(expensive_flag)
print(f'True: {expensive_flag[expensive_flag].size}\nFalse: {expensive_flag[~expensive_flag].size}')



2
price_per_m2_array mean: 948766.77, median: 791046.72, min: 151609.63, max: 3270821.84

3
Характеристика квартиры у которой max price per m2: price: 118.60, area: 36.26, rooms: 2

4
QTY of apartmanets vy m2 > avg: 77

5and6
[ True False False False  True False False False False False False False
 False False False False False False False  True  True False False False
  True False  True False False False False False False  True False False
 False False False  True False  True  True False  True  True  True False
 False False False False False False False False  True  True False False
  True False False False False False  True False False  True  True False
 False False False False  True False False False False  True  True False
 False  True False  True False  True False False False False False False
  True False False False False  True  True False  True  True False  True
 False  True False False False  True False False False  True False  True
 False False False False False  True False 

### 1.5. Нормализация и краткий вывод — 4 балла

1. Нормализуйте `area_array` и `price_array` методом Min-Max в диапазон [0, 1]. Формулу реализуйте самостоятельно средствами NumPy.
2. Проверьте минимальное и максимальное значения нормализованных массивов.
3. Выведите первые 5 исходных и первые 5 нормализованных значений.
4. Напишите 2–3 предложения: зачем нужна нормализация и в каких задачах она может быть полезна.

In [18]:
area_normalized = (area_array - area_array.min()) / (area_array.max() - area_array.min())
price_normalized = (price_array - price_array.min()) / (price_array.max() - price_array.min())

print(f'area normalized min: {area_normalized.min():.2f}, max: {area_normalized.max():.2f}')
print(f'price normalized min: {price_normalized.min():.2f}, max: {price_normalized.max():.2f}')

print("\narea original:", area_array[:5])
print("area normalized:", area_normalized[:5])

print("\nprice original:", price_array[:5])
print("price normalized:", price_normalized[:5])

area normalized min: 0.00, max: 1.00
price normalized min: 0.00, max: 1.00

area original: [121.74  79.86 132.32 112.17  36.77]
area normalized: [0.7782019  0.43807358 0.86412735 0.70047917 0.08811825]

price original: [ 96.79  29.13  71.29  68.99 105.05]
price normalized: [0.77841671 0.12995975 0.53402339 0.51198007 0.85758099]


Нормализация приводит площадь и цену к общему диапазону от 0 до 1, сохраняя порядок значений. Она полезна в машинном обучении, например при кластеризации и поиске ближайших соседей, чтобы признаки с большими числовыми значениями не доминировали при расчёте расстояний.

## Задание 2. NumPy: многомерные массивы и анализ тензора (25 баллов)

**Контекст:** имеются почасовые температуры для 3 городов за 7 дней. Данные представлены трёхмерным массивом NumPy.

### 2.1. Создание и исследование тензора — 4 балла

1. Создайте тензор `T` формы `(7, 24, 3)`: ось 0 — дни, ось 1 — часы, ось 2 — города.
2. Заполните тензор случайными температурами от -10 до +35 °C.
3. Выведите `shape`, `ndim`, `size` и `dtype`.
4. Отдельно объясните, что обозначает каждая ось и чему равен `T.shape[0]`, `T.shape[1]`, `T.shape[2]`.

In [19]:
import numpy as np
rng = np.random.default_rng(42)
T = rng.integers(-10, 36, size=(7, 24, 3))
print(f'Shape: {T.shape} \nndim: {T.ndim} \nsize: {T.size} \ndtype: {T.dtype}')
# print(T)

Shape: (7, 24, 3) 
ndim: 3 
size: 504 
dtype: int64


### оси тензора

Тензор `T` имеет форму `(7, 24, 3)`:

- **Ось 0 — дни:** `T.shape[0] = 7`, то есть 7 дней.
- **Ось 1 — часы:** `T.shape[1] = 24`, то есть 24 часа в каждом дне.
- **Ось 2 — города:** `T.shape[2] = 3`, то есть 3 города.

Например, `T[0, 12, 2]` — температура в первый день, в 12:00, в третьем городе.


### 2.2. Индексация и срезы — 6 баллов

1. Выведите температуры второго города за все дни и все часы. Укажите форму полученного массива.


> **Подсказка:** На защите будьте готовы объяснить, почему `T[:, :, 1]`, `T[0, :, :]` и `T[:, 8, :]` имеют разные формы.

In [20]:
print(f'All temperatures of city 2:\n {T[:, :, 1]}')
print('Shape:', T[:, :, 1].shape)

All temperatures of city 2:
 [[ 25   9  22  14  25  13  10  -2  19  15   0  30  28  -3   6  10  25   6
   -8  24  24   8   7  26]
 [ -5   5  20  -3  22  25  27  31  21  28 -10  25  22  15  16   1  20  29
   25  16  15  17  10   2]
 [ 29  27   2   9  15  35   8   4  -9  25  11  31  -3  12   7  21   6   5
   34  12   2   2  22  23]
 [ -6  -5  -1  27  -2 -10  23   4  16  18   9  -2   5  21  17   3  16  18
   -9   9  26   2  12  -3]
 [ 11  33  13   3  28  14  15  20   2  23  25  18   0  -4   7  11   4  19
   13  35  -3   3  35  -6]
 [ 30  -5   4   5   7  23   2  33  -5  31  -2  23  -1  35  34   9 -10   2
   21   3  21   9  26  -7]
 [ -4  16   8  -4  20  13  32  17  19  23   9  -3  32   9  26  20  28  11
   14  32  17  -5   9  34]]
Shape: (7, 24)


2. Выведите все данные первого дня для трёх городов. Укажите форму.

In [21]:
print(f'All temperatures of day 1 for all cities:\n {T[0, :, :]}')
print('Shape:', T[0, :, :].shape)

All temperatures of day 1 for all cities:
 [[-6 25 20]
 [10  9 29]
 [-7 22 -1]
 [-6 14 34]
 [23 25 23]
 [26 13 -5]
 [28 10 13]
 [ 7 -2 32]
 [25 19  8]
 [27 15 10]
 [10  0 -6]
 [15 30 -8]
 [29 28  2]
 [19 -3 24]
 [22  6 -7]
 [34 10 31]
 [21 25 24]
 [-2  6 11]
 [12 -8 15]
 [-3 24 21]
 [32 24  6]
 [34  8  4]
 [31  7 -7]
 [11 26 -2]]
Shape: (24, 3)


3. Получите одно значение: температура в 3-й день, в 12:00, во 2-м городе. Обратите внимание на индексацию с нуля.

In [22]:
print(f'temperature in 3rd day at 12:00 in 2nd city:\n {T[2, 12, 1]}')

temperature in 3rd day at 12:00 in 2nd city:
 -3


4. Выведите температуры всех городов с 09:00 до 18:00 за всю неделю.

In [23]:
print(f'Temperatures of all cities from 09:00 to 18:00 for the entire week:\n {T[:, 9:19, :]}')

Temperatures of all cities from 09:00 to 18:00 for the entire week:
 [[[ 27  15  10]
  [ 10   0  -6]
  [ 15  30  -8]
  [ 29  28   2]
  [ 19  -3  24]
  [ 22   6  -7]
  [ 34  10  31]
  [ 21  25  24]
  [ -2   6  11]
  [ 12  -8  15]]

 [[ -4  28  -1]
  [ 27 -10  26]
  [ 26  25  20]
  [ 11  22   2]
  [ 25  15  11]
  [ 13  16  -9]
  [ -4   1  -5]
  [ 10  20  20]
  [ 11  29  16]
  [ -7  25  16]]

 [[ -6  25  23]
  [ 22  11  22]
  [ -3  31  13]
  [ 33  -3  12]
  [ 22  12  10]
  [ -3   7   0]
  [  3  21  18]
  [ 17   6  34]
  [ -6   5  -5]
  [  5  34   6]]

 [[ 19  18  -7]
  [ 24   9  26]
  [ -9  -2  12]
  [ -2   5  20]
  [ -4  21  -6]
  [ -2  17  26]
  [ -3   3  32]
  [ 35  16  10]
  [  5  18  17]
  [  3  -9  -4]]

 [[ 20  23  23]
  [ 33  25   4]
  [ -6  18  32]
  [ 34   0   0]
  [ -9  -4  15]
  [ 15   7   4]
  [ 28  11  27]
  [ 34   4  -1]
  [ 33  19   3]
  [ 18  13  31]]

 [[ 28  31  -3]
  [ 24  -2   6]
  [ 17  23  30]
  [ -8  -1   2]
  [  4  35  25]
  [  4  34  12]
  [ 13   9  -4]
  [ 32 -1

5. Выведите температуры первого города за выходные дни (6-й и 7-й дни).

In [24]:
print(f'Temperatures of the first city over the weekend (6th and 7th days):\n {T[5:7, :, 0]}')

Temperatures of the first city over the weekend (6th and 7th days):
 [[ 28  31  11  25   8  -6   7  33  -9  28  24  17  -8   4   4  13  32   0
  -10  -5  32  16  29  22]
 [ 27  -5  -7   3  13  33  34  -9  -4  -6  24  34  -2  27  19 -10   3  26
    5  17  24   8  32  19]]


6. Получите массив температур всех городов ровно в 08:00 за 7 дней и укажите его форму.

In [25]:
print(f'The temperatures of all cities at exactly 08:00 over a period of 7 days:\n {T[:, 8, :]}')
print(f'Shape: {T[:, 8, :].shape}')

The temperatures of all cities at exactly 08:00 over a period of 7 days:
 [[25 19  8]
 [ 1 21 19]
 [ 5 -9 -6]
 [-4 16 -6]
 [-5  2  1]
 [-9 -5 -7]
 [-4 19 28]]
Shape: (7, 3)


### 2.3. Агрегирование по осям — 7 баллов

> **Подсказка:** В этом разделе важно правильно выбирать `axis`. В комментариях укажите, какие оси вы сворачиваете при каждом вычислении.

1. `axis=(0, 1)` — сворачиваем дни и часы, оставляем города.
2. `axis=(0, 1)` — сворачиваем дни и часы, оставляем города.
3. `axis=(1, 2)` — сворачиваем часы и города, оставляем дни.
4. `axis=1` — сворачиваем часы, оставляем дни и города.
5. `axis=(0, 2)` — сворачиваем дни и города, оставляем часы.
6. `axis=(1, 2)` — сворачиваем часы и города, оставляем дни.
7. `axis=(0, 1)` — сворачиваем дни и часы, оставляем города.






1. Найдите среднюю температуру по каждому городу за всю неделю.


In [26]:
avg_temp_by_city = T.mean(axis=(0, 1))
print('AVG temp by cities', avg_temp_by_city)

AVG temp by cities [12.69047619 13.125      11.75595238]


2. Найдите минимальную и максимальную температуру для каждого города.

In [27]:
min_temp_by_city = T.min(axis=(0, 1))
max_temp_by_city = T.max(axis=(0, 1))

print('MIN temp by city' , min_temp_by_city)
print('MAX temp by city' , max_temp_by_city)

MIN temp by city [-10 -10 -10]
MAX temp by city [35 35 35]


3. Для каждого дня найдите максимальную температуру среди всех часов и городов.

In [28]:
max_temp_by_day = T.max(axis=(1, 2))

print('MAX temp by day' , max_temp_by_day)

MAX temp by day [34 35 35 35 35 35 34]


4. Для каждого дня вычислите среднюю температуру отдельно по каждому городу. Результат должен иметь форму `(7, 3)`.

In [29]:
avg_temp_by_day_city = T.mean(axis=(1))

print('AVG temp by day and city' , avg_temp_by_day_city)

AVG temp by day and city [[16.33333333 13.875      11.29166667]
 [ 8.95833333 15.58333333 13.54166667]
 [12.5        13.75       12.54166667]
 [11.16666667  7.83333333 12.25      ]
 [13.58333333 13.29166667 11.54166667]
 [13.58333333 12.          8.70833333]
 [12.70833333 15.54166667 12.41666667]]


5. Определите час суток с минимальной средней температурой, усреднив значения по всем дням и городам.

In [30]:
avg_temp_by_hour = T.mean(axis=(0, 2))
coldest_hour = avg_temp_by_hour.argmin()
print('Coldest hour' , coldest_hour)

Coldest hour 8


6. Определите день с самой высокой средней температурой.

In [31]:
avg_temp_by_day = T.mean(axis= (1, 2))
hottest_day = avg_temp_by_day.argmax()
print('Hottest day' , hottest_day+1)

Hottest day 1


7. Определите город с самой высокой средней температурой за неделю.

In [32]:
hottest_week_city = avg_temp_by_city.argmax()
print('hottest_week_city', hottest_week_city+1)

hottest_week_city 2


### 2.4. Изменение формы и проверка данных — 4 балла


1. Преобразуйте `T` из формы `(7, 24, 3)` в матрицу формы `(7, 72)`.
2. Объясните, что теперь находится в одной строке новой матрицы.

каждай строка теперь представляет 1 день но содержит по 72 значения по часам и городам напрмер первые три значения температуры в 3х городах в 0 часво итд

In [33]:
T_matrix = T.reshape(7,72)
print(T_matrix) 

[[ -6  25  20  10   9  29  -7  22  -1  -6  14  34  23  25  23  26  13  -5
   28  10  13   7  -2  32  25  19   8  27  15  10  10   0  -6  15  30  -8
   29  28   2  19  -3  24  22   6  -7  34  10  31  21  25  24  -2   6  11
   12  -8  15  -3  24  21  32  24   6  34   8   4  31   7  -7  11  26  -2]
 [ 11  -5  21  11   5   0  15  20  33  10  -3  28  18  22  -6   4  25  28
   10  27  28   7  31   3   1  21  19  -4  28  -1  27 -10  26  26  25  20
   11  22   2  25  15  11  13  16  -9  -4   1  -5  10  20  20  11  29  16
   -7  25  16  19  16  15  -6  15  26   3  17  -9   5  10  35  -1   2   8]
 [ 35  29  -9   0  27  -8  29   2  32   3   9  20  -5  15  13  26  35  20
    8   8   9  27   4  -3   5  -9  -6  -6  25  23  22  11  22  -3  31  13
   33  -3  12  22  12  10  -3   7   0   3  21  18  17   6  34  -6   5  -5
    5  34   6  31  12  22  11   2  25  34   2  25   2  22  26  10  23   2]
 [ -7  -6  10  31  -5  10  22  -1  23   4  27  16  15  -2  11  29 -10  24
   12  23  20   9   4  18  -4  16  

3. Верните матрицу обратно к форме `(7, 24, 3)`.
4. Проверьте, совпадают ли исходный и восстановленный тензоры с помощью `np.array_equal()` или `np.allclose()`.

In [34]:
T_restored = T_matrix.reshape(7, 24, 3)

print('Совпадают:', np.array_equal(T, T_restored)) 

Совпадают: True


### 2.5. Сравнение городов и интерпретация — 4 балла

1. Преобразуйте данные каждого города в последовательность из 168 температур (7 × 24).
2. Постройте корреляционную матрицу 3 × 3 с помощью `np.corrcoef`.
3. Не учитывая диагональ, определите пару городов с наибольшим коэффициентом корреляции.
4. Напишите 2–3 предложения, что означает положительная, отрицательная и близкая к нулю корреляция в контексте температур.

In [35]:
T_city = T.reshape(168, 3).T
corr_maytix = np.corrcoef(T_city)
print('Correlation matrix:\n', corr_maytix)
rows, cols = np.triu_indices(3, k=1)
most_corr_idx = corr_maytix[rows, cols].argmax()
city_1 = rows[most_corr_idx]
city_2 = cols[most_corr_idx]
print('Cities:', city_1 + 1, city_2 + 1)
print('Correlation:', corr_maytix[city_1, city_2])

Correlation matrix:
 [[ 1.         -0.05829554 -0.09031105]
 [-0.05829554  1.          0.00340061]
 [-0.09031105  0.00340061  1.        ]]
Cities: 2 3
Correlation: 0.0034006110500383133


Положительная корреляция означает, что температуры двух городов обычно повышаются и понижаются вместе. Отрицательная — что температура в одном городе повышается, а в другом обычно понижается. Близкая к нулю корреляция означает, что нет выраженной линейной связи: повышение или понижение температуры в одном городе не сопровождается определённым изменением в другом.

## Задание 3. Pandas: загрузка, очистка и подготовка данных (25 баллов)

**Контекст:** работа с датасетом объявлений о продаже квартир. Используйте CSV-файлы или единый CSV-файл, выданный преподавателем.

### 3.1. Загрузка и первичное исследование — 4 балла

1. Импортируйте pandas, numpy и необходимые библиотеки для визуализации.
2. Если преподаватель выдал несколько CSV-файлов, прочитайте их и объедините в один DataFrame с помощью `pd.concat(..., ignore_index=True)`. Если выдан один файл — просто загрузите его.
3. Выведите первые 5 и последние 5 строк DataFrame.
4. Выведите `df.shape`, `df.columns`, `df.dtypes`, `df.info()` и `df.describe(include="all")`.
5. В 2–3 предложениях опишите размер датасета и типы содержащихся признаков.

In [41]:
import pandas as pd
import numpy as np
df = pd.read_csv('apartments_source.csv')
print(f'первые 5: \n{df.head()}\n последние 5: \n{df.tail()}')

первые 5: 
  Advertisement_ID Publication_Date               Apartment_Address  \
0         APT-0458       2026-08-11  Учебная улица 1, дом 31, кв. 8   
1         APT-0082       2026-08-08   Учебная улица 1, дом 6, кв. 7   
2         APT-0170       2026-09-04  Учебная улица 6, дом 12, кв. 5   
3         APT-0211       2026-07-20  Учебная улица 5, дом 15, кв. 1   
4         APT-0309       2026-07-12  Учебная улица 4, дом 21, кв. 9   

  City_District Residential_Complex_Name  Number_of_Rooms Apartment_Area_m2  \
0         Есиль           Учебный ЖК 1-2                3              89.8   
1         Есиль           Учебный ЖК 1-1                2              56.0   
2      Сарайшық                      NaN                3              70.9   
3      Байконыр                      NaN                2              57.6   
4      Сарыарка                      NaN                1           28.4 м²   

  Apartment_Price_KZT  Apartment_Floor  Total_Building_Floors  \
0            61700000 

In [42]:
print(f'Shape: {df.shape}')
print(f'Columns:\n{df.columns}')
print(f'Dtypes:\n{df.dtypes}')

Shape: (500, 12)
Columns:
Index(['Advertisement_ID', 'Publication_Date', 'Apartment_Address',
       'City_District', 'Residential_Complex_Name', 'Number_of_Rooms',
       'Apartment_Area_m2', 'Apartment_Price_KZT', 'Apartment_Floor',
       'Total_Building_Floors', 'Building_Construction_Year',
       'Apartment_Condition'],
      dtype='str')
Dtypes:
Advertisement_ID                  str
Publication_Date                  str
Apartment_Address                 str
City_District                     str
Residential_Complex_Name          str
Number_of_Rooms                 int64
Apartment_Area_m2                 str
Apartment_Price_KZT               str
Apartment_Floor                 int64
Total_Building_Floors           int64
Building_Construction_Year    float64
Apartment_Condition               str
dtype: object


In [43]:
print('\nInfo:')
df.info()


Info:
<class 'pandas.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 12 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Advertisement_ID            500 non-null    str    
 1   Publication_Date            500 non-null    str    
 2   Apartment_Address           500 non-null    str    
 3   City_District               500 non-null    str    
 4   Residential_Complex_Name    413 non-null    str    
 5   Number_of_Rooms             500 non-null    int64  
 6   Apartment_Area_m2           500 non-null    str    
 7   Apartment_Price_KZT         500 non-null    str    
 8   Apartment_Floor             500 non-null    int64  
 9   Total_Building_Floors       500 non-null    int64  
 10  Building_Construction_Year  470 non-null    float64
 11  Apartment_Condition         475 non-null    str    
dtypes: float64(1), int64(3), str(8)
memory usage: 47.0 KB


In [44]:
print(f'\nDescribe:\n{df.describe(include="all")}')


Describe:
       Advertisement_ID Publication_Date               Apartment_Address  \
count               500              500                             500   
unique              490               92                             480   
top            APT-0002       2026-07-12  Учебная улица 5, дом 1, кв. 14   
freq                  2               12                               2   
mean                NaN              NaN                             NaN   
std                 NaN              NaN                             NaN   
min                 NaN              NaN                             NaN   
25%                 NaN              NaN                             NaN   
50%                 NaN              NaN                             NaN   
75%                 NaN              NaN                             NaN   
max                 NaN              NaN                             NaN   

       City_District Residential_Complex_Name  Number_of_Rooms  \
count     

In [46]:
print(f'data types: \n{df.dtypes}')

data types: 
Advertisement_ID                  str
Publication_Date                  str
Apartment_Address                 str
City_District                     str
Residential_Complex_Name          str
Number_of_Rooms                 int64
Apartment_Area_m2                 str
Apartment_Price_KZT               str
Apartment_Floor                 int64
Total_Building_Floors           int64
Building_Construction_Year    float64
Apartment_Condition               str
dtype: object


Датасет содержит 500 строк и 12 столбцов, описывающих объявления о продаже квартир. Числовые признаки — это Number_of_Rooms, Building_Construction_Year, Total_Building_Floors и Apartment_Floor, а текстовые — Advertisement_ID, Publication_Date, Apartment_Address, City_District и остальные. Также Apartment_Area_m2 и Apartment_Price_KZT пока хранятся как строки, так как некоторые значения содержат единицы измерения, например 56.5 м², или знак тенге.

### 3.2. Проверка и обработка дубликатов — 4 балла

1. Посчитайте количество полных дубликатов строк с помощью `duplicated().sum()`.
2. Выведите несколько найденных дубликатов, если они существуют.


In [50]:
duplicates_all = df.duplicated().sum()
print('QTY of full duplicates', duplicates_all)
display(df[df.duplicated()].head())

QTY of full duplicates 10


,Advertisement_ID,Publication_Date,Apartment_Address,City_District,Residential_Complex_Name,Number_of_Rooms,Apartment_Area_m2,Apartment_Price_KZT,Apartment_Floor,Total_Building_Floors,Building_Construction_Year,Apartment_Condition
115,APT-0007,2026-07-10,"Учебная улица 3, дом 1, кв. 7",Алматы,Учебный ЖК 3-3,2,62.4,31400000,5,9,1997.0,Евроремонт
251,APT-0010,2026-08-28,"Учебная улица 5, дом 1, кв. 10",Байконыр,NaN,3,96.2,42500000,8,9,1995.0,Хорошее
364,APT-0001,2026-07-12,"Учебная улица 6, дом 1, кв. 1",Сарайшық,Учебный ЖК 6-3,1,35.7 м²,18000000,4,22,1983.0,Хорошее
392,APT-0005,2026-09-12,"Учебная улица 4, дом 1, кв. 5",Сарыарка,Учебный ЖК 4-1,2,67.7,28900000,11,12,2010.0,Хорошее
393,APT-0002,2026-07-26,"Учебная улица 4, дом 1, кв. 2",Сарыарка,Учебный ЖК 4-3,1,30.0,9500000,4,9,1980.0,Хорошее


3. Проверьте дубликаты по логически важным колонкам. Например: `ID + Price`, `Address + Price` или другая комбинация, соответствующая вашему датасету.
4. Примите решение: нужно ли удалять дубликаты. Обязательно объясните решение.
5. Если дубликаты удаляются, покажите количество строк до и после удаления.


In [53]:
key_columns = ['Apartment_Address', 'Apartment_Price_KZT']

duplicates_mask = df.duplicated(subset=key_columns, keep=False)

display(df[duplicates_mask].sort_values(key_columns).head(6))
print(f'Количество повторов: {df.duplicated(subset=key_columns).sum()}')

,Advertisement_ID,Publication_Date,Apartment_Address,City_District,Residential_Complex_Name,Number_of_Rooms,Apartment_Area_m2,Apartment_Price_KZT,Apartment_Floor,Total_Building_Floors,Building_Construction_Year,Apartment_Condition
119,APT-0011,2026-07-29,"Учебная улица 3, дом 1, кв. 11",Алматы,Учебный ЖК 3-1,1,35.7,17400000,14,18,2022.0,Евроремонт
353,APT-0481,2026-07-29,"Учебная улица 3, дом 1, кв. 11",Алматы,Учебный ЖК 3-1,1,35.7,17400000,14,18,2022.0,Евроремонт
265,APT-0004,2026-07-06,"Учебная улица 3, дом 1, кв. 4",Алматы,Учебный ЖК 3-3,1,40.2,19900000,3,5,1999.0,NaN
488,APT-0004,2026-07-06,"Учебная улица 3, дом 1, кв. 4",Алматы,Учебный ЖК 3-3,1,40.2,19900000,3,5,1999.0,NaN
79,APT-0007,2026-07-10,"Учебная улица 3, дом 1, кв. 7",Алматы,Учебный ЖК 3-3,2,62.4,31400000,5,9,1997.0,Евроремонт
115,APT-0007,2026-07-10,"Учебная улица 3, дом 1, кв. 7",Алматы,Учебный ЖК 3-3,2,62.4,31400000,5,9,1997.0,Евроремонт


Количество повторов: 20


Полные дубликаты стоит удалить, поскольку они повторяют одни и те же объявления. Записи с одинаковым точным адресом, включая номер квартиры, ценой и остальными характеристиками также считаю повторными: возможно, объявление было добавлено дважды по ошибке. Если цена или другие данные различаются, такие записи лучше сохранить, поскольку одну квартиру могли выставить и хозяин, и риелтор.

In [54]:
rows_before = len(df)

df = df.drop_duplicates(
    subset=key_columns,
    keep='first'
).reset_index(drop=True)

print(f'Строк до удаления: {rows_before} Строк после удаления: {len(df)} Удалено: {rows_before - len(df)}')

Строк до удаления: 500 Строк после удаления: 480 Удалено: 20
